scripts/rouleth/rouleth.ts

```diff
@@ -0,0 +1,210 @@
+          addressArray.push({id: elem.id, nb: elem.numLands})
+        );
+      }
+      resElemNb = data[field].length;
+      if (resElemNb > 0) {
+        const newLastId = data[field][resElemNb - 1].id;
+        if (lastId === newLastId) {
+          console.log('same query, stop');
+          break;
+        }
+        lastId = newLastId;
+      }
+    }
+    return addressArray;
+  };
+
+  return await query(queryString, 'owners', {blockNb});
+};
+
+const getAddressFromBack = async (address: string[]): Promise<string[]> => {
+  const provider = ethers.provider;
+
+  // check if address is a contract
+  const isContract = async (address: string) => {
+    const result = await provider.getCode(address);
+    return result !== '0x';
+  };
+
+  // remove whitespaces, check if address is valid, resolve ens
+  const verifyAddress = async (address: string) => {
+    const addressClean: string = address.replace(/\s/g, '');
+    if (ethers.utils.isAddress(addressClean)) {
+      if (await isContract(addressClean)) {
+        console.error('contract are not eligible:\t', addressClean);
+        return false;
+      }
+      return addressClean;
+    } else {
+      try {
+        const resolvedEns = await provider.resolveName(addressClean);
+        if (resolvedEns !== null && address !== '.eth') {
+          return resolvedEns;
+        }
+        console.error('failed to resolve ens:\t\t', addressClean);
+        return false;
+      } catch (err) {
+        console.error(err);
+      }
+    }
+  };
+
+  // check all address are valid, and await resolution
+  const promiseResolver = await Promise.all(
+    address.map((add) => verifyAddress(add))
+  );
+
+  // insure we only have string in the array and false / undefined values from address check are removed
+  const addressFromBack: string[] = promiseResolver.filter(
+    (val) => typeof val === 'string'
+  ) as string[];
+
+  return addressFromBack;
+};
+
+const lottery = async (
+  addBack: Array<string>,
+  addGraph: Array<LandOwner>,
+  maxWinnerNb: number,
+  blockNumber: number
+) => {
+  const seed: string = (await ethers.provider.getBlock(blockNumber)).hash;
+
+  class SeededRand {
+    private seed;
+    constructor(seed: string) {
+      this.seed = seed;
+    }
+
+    nextInteger(): number {
+      this.seed = solidityKeccak256(['bytes32'], [this.seed]);
+      return parseInt(this.seed, 16) % Math.floor(Math.random() * 100);
+      // return BigNumber.from(this.seed).mod(Number.MAX_SAFE_INTEGER).toNumber();
+    }
+  }
+
+  // return an array of address subscribe on the back
+  // if an address own x land, add x time the address to the array (x >= 1)
+  const extendedAddressArray = async () => {
+    const extendedArray: Array<string> = [];
+    addBack.filter((itemAddress: string) => {
+      extendedArray.push(itemAddress);
+      addGraph.map((itemLandOwner: LandOwner) => {
+        if (itemLandOwner.id === itemAddress) {
+          for (let i = 0; i < itemLandOwner.nb; i++) {
```

Comment:
actually there is another design choice :
we need to add the address n +1 times (n being the number of land owned)